### DISCUSSION TITLE
Atomic transactions in a web framework

### DISCUSSION THREAD

**Discussion by peterswords** [piccolo-orm/piccolo#185] [Q&A]
How would you wrap a fastapi function in a transaction and have it roll back if there are any errors during execution? I looked at transactions in the documentation, and see the "async with ..._meta.db.transaction()" approach. So I could write the "async with" at the top of every path function, but is there a way to make this happen generically (e.g. a fastapi dependency).

**Comment by dantownsend** [comment]
It's a good question.

I tend to manually wrap my view logic in the transaction context manager when required:

```python
@app.get("/")
async def my_endpoint():
   try:
        async with MyTable._meta.db.transaction():
            # a bunch of queries
   except Exception:
        return JsonResponse({"message": "something went wrong"})

    return JsonResponse({"message": "success"})
```

It would be nice to move the transaction handling out of the view though. A decorator is probably the best way to achieve this. Something like:

```python
@app.get("/")
@piccolo_transaction
async def my_endpoint():
    ...
```

There isn't an out of the box solution for this yet, but I've created a ticket for it.

https://github.com/piccolo-orm/piccolo_api/issues/75

**Reply by peterswords** [reply]
That's great. Once it can be done in principle I'm not too concerned about an OOTB solution. 

Off topic: Many thanks, wasn't expecting a fast response on a Sunday! I came across Piccolo on your recent chat on the "Talk Python to Me" podcast. I've been kicking the tyres since and love the look of it. I have some wider architectural questions, but don't want to pollute your discussion board. Is it ok to air them under the "General" discussions category?

**Reply by dantownsend** [reply]
Of course - feel free to leave any feedback you have!
